makefile:add initial documentation and clean up aliases (#43)
content-services/content-sources-backend: mk/README.md

# Makefile composition

The repository is using a way to compose the `Makefile` file which
empower the single responsability principal taken into the Makefile
file usage.

> Very quickly, we have the following files:

Reviewer: ```suggestion
We have the following files:
```
Try not to make judgements about things (brief, easy, quick etc.)

```raw
Makefile                    # Minimal main Makefile
mk
├── README.md               # This documentation file
├── db.mk
├── docker.mk
├── go-rules.mk

> ├── help.mk                 # Rule to print out generated help content from the own Makefile files

Reviewer: ```suggestion
├── help.mk                 # Rule to print out generated help content from the Makefile's files
```

├── includes.mk             # Entry point included into the main Makefile
├── meta-db.mk
├── meta-docker.mk
├── meta-general.mk
├── plantuml.mk
├── printvars.mk            # print out variables
├── projectdir.mk           # Store at PROJECT_DIR the base directory for the repository
├── swag.mk
└── variables.mk            # Default values to the project variables that has not been
                            # overrided by the environment variables nor by configs/config.yaml
                            # file
```

> ## Quickly usage

Reviewer: ```suggestion
## Usage
```

- Prepare database without the migrations: `make db-clean db-up RUN_MIGRATION=`
- Open the postgres cli in the current executing container: `make db-cli-connect`
- Populate database with random records: `make db-migrate-seed`
- Build the docker container image: `make docker-build`
- Push the docker image to the registry:

  ```sh
  make docker-build docker-push
  ```

  > By default it uses the local username as the `QUAY_USER`. Can be overrided.
  > It uses the image name that is published by default.
  > It tag the image with the current git short hash.
  > We can publish in our personal quay.io account, deploy in ephemeral
  > environment the fresh image container, and see if the image behave
  > as we expect before to push our changes to the repository.

  ```sh
  make docker-build docker-push DOCKER_IMAGE="quay.io/my-user/my-image:my-tag"
  ```

  ```sh
  make docker-build docker-push QUAY_USER="my-user"
  ```

  ```sh
  make docker-build docker-push DOCKER_IMAGE_BASE="quay.io/my-user/my-image"
  ```

  > See `make printvars` to check related variables.

## More detailed explanation

It uses single responsability principal and split basic or logical groupable

> rules in a single files.

Reviewer: ```suggestion
rules in a single file.
```

We could group into the following way:

@Minimal and layout composition

- `Makefile`: Entrypoint and the only responsability is to include
  the list of includes and set the dafault rule.
- `mk/includes.mk`: This define the composition for your repository
  and it reference other `mk/*` files, and the `mk/meta-*` files
  which allow the `help` rule to print out the group title in the
  right position.

@Variables

> - `mk/projectdir.mk`: It uses to be the first include and just store

Reviewer: ```suggestion
- `mk/projectdir.mk`: It uses the first include and stores
```

  at the `PROJECT_DIR` variable the path to the repository in the
  file system.
- `mk/variables.mk`: It defines the default values for variables.

@Help infrastructure

> - `mk/meta-*.mk`: This files is just one line content each one,

Reviewer: ```suggestion
- `mk/meta-*.mk`: This file is just one line of content for each one,
```

  but that allow us to print the group label into the help generated
  as we want. It can not be written directly into the `mk/includes.mk`
  file because of the order the files are processed when the `help`
  rule is invoked.

> - `mk/help.mk`: Contain the rule that generate the help content.

Reviewer: ```suggestion
- `mk/help.mk`: Contains the rule that generates the help content.
```

@Rules

> - `mk/go-rules.mk`: Provide useful rules to build/clean/test

Reviewer: ```suggestion
- `mk/go-rules.mk`: Provide useful rules to build, clean, and test
```
No slash in text please

  `build` rule I will get the `release/nice-command` binary
  if the compilation is succesful, and no changes are needed
  to the Makefile if a new `cmd/other-nice-command/main.go`
  is created. "Write once, use forever".
- `mk/docker.mk`: Generic rules to build and push container images.
  It has an interesting set of variables that combined properly
  allow us a lot of customization and create new rules on them.
- `mk/db.mk`: Rules to start/stop a postgres container database
  and clean the data volume. It adds some other rules to quickly
  invoke the migration and seed the database with random values.
- `mk/plantuml.mk`: Invoke plantuml to generate diagrams in SVG
  format from the `docs/*.puml` files. It requires that plantuml
  is installed in your environment. Add a generic rule for it, so
  that no changes are needed when new `.puml` file is added to the
  `doc/` directory.

@Miscelanea

- `mk/printvars.mk`: It helps to fix wrong behaviors so it print out
  the variable name and the value not evaluated, so we can realize
  what final values could be applied and how should the values be
  expanded.

## Contributing

> Do you have a rule or some set of rules that could be useful here? please

Reviewer: ```suggestion
Do you have a rule or some set of rules that could be useful here?
```
No need for "please" in techincal docs.

> do not hesitate to contribute to add and review them together. It does

Reviewer: ```suggestion
Do not hesitate to contribute and review them together. It does
```

# Makefile composition

The repository is using a way to compose the `Makefile` file which
empower the single responsability principal taken into the Makefile
file usage.

Very quickly, we have the following files:

```raw
Makefile                    # Minimal main Makefile
mk
├── README.md               # This documentation file
├── alias.mk
├── db.mk
├── docker.mk
├── go-rules.mk
├── help.mk                 # Rule to print out generated help content from the own Makefile files
├── includes.mk             # Entry point included into the main Makefile
├── meta-alias.mk           # Group title included from the mk/includes.mk file
├── meta-db.mk
├── meta-docker.mk
├── meta-general.mk
├── plantuml.mk

> ├── printvars.mk            # print the list of defined variables, without evaluate the value

Author: thanks for this; this needs some clarification; I have checked out, and not always is printing out the evaluated value. Only when the last assignment was using the `:=` and `?=` operators, else it prints out the expression, instead the evaluated value.

Example:

For this `makefile` file included which define some variables in different ways:

```makefile
EXAMPLE_01 := example_01
EXAMPLE_02 = example_02
EXAMPLE_03 += example_03
EXAMPLE_04 ?= example_04

EXAMPLE_05 = $(EXAMPLE_01) example_05
EXAMPLE_06 = $(EXAMPLE_02) example_06
EXAMPLE_07 = $(EXAMPLE_03) example_07
EXAMPLE_08 = $(EXAMPLE_04) example_08

EXAMPLE_09 += $(EXAMPLE_01) example_09
EXAMPLE_10 += $(EXAMPLE_02) example_10
EXAMPLE_11 += $(EXAMPLE_03) example_11
EXAMPLE_12 += $(EXAMPLE_04) example_12

EXAMPLE_01 := $(EXAMPLE_01)
EXAMPLE_02 = $(EXAMPLE_02)
EXAMPLE_03 += $(EXAMPLE_03)
EXAMPLE_04 ?= $(EXAMPLE_04)

```

The resulting output for `make printvars`

```raw
EXAMPLE_01=example_01
EXAMPLE_02=$(EXAMPLE_02)
EXAMPLE_03=example_03 $(EXAMPLE_03)
EXAMPLE_04=example_04
EXAMPLE_05=$(EXAMPLE_01) example_05
EXAMPLE_06=$(EXAMPLE_02) example_06
EXAMPLE_07=$(EXAMPLE_03) example_07
EXAMPLE_08=$(EXAMPLE_04) example_08
EXAMPLE_09=$(EXAMPLE_01) example_09
EXAMPLE_10=$(EXAMPLE_02) example_10
EXAMPLE_11=$(EXAMPLE_03) example_11
EXAMPLE_12=$(EXAMPLE_04) example_12
```

> ├── printvars.mk            # print the list of defined variables, without evaluate the value

Author: I think is better just to say `print out the variable values`, if that values are evaluated or not, depends on which operator was used the last and if it is including into the value some substitution or calling some command inside the variable value.

> ├── printvars.mk            # print the list of defined variables, without evaluate the value

Author: slight update into the comment about 'printvars.mk' file.

  ```

  ```sh
  make docker-build docker-push DOCKER_IMAGE_BASE="quay.io/my-user/my-image"
  ```

  > See `make printvars` to check related variables.

## More detailed explanation

It uses single responsability principal and split basic or logical groupable
rules in a single files.

We could group into the following way:

@Minimal and layout composition

- `Makefile`: Entrypoint and the only responsability is to include
  the list of includes and set the dafault rule.
- `mk/includes.mk`: This define the composition for your repository
  and it reference other `mk/*` files, and the `mk/meta-*` files
  which allow the `help` rule to print out the group title in the
  right position.

@Variables

- `mk/projectdir.mk`: It uses to be the first include and just store
  at the `PROJECT_DIR` variable the path to the repository in the
  file system.
- `mk/private.mk`: Not used into this project, but it could define
  by using conditional assignment values that we want to set for
  our make rules; useful for the specific rules related with our
  tech-stack, or when building container images, to customize
  our own registry (quay.io/my-username); provide a mechanism

>   of customization to avoid repeat steps manually as `export MY_VAR=my-value`.

Reviewer 2: I think we should either remove the text about private.mk or add support for it in in our project (likely adding it to includes.mk, adding a blank one, and adding it to git ignore?)

Author: Thanks! Updated.

Author: thanks! I removed the references to `private.mk`; I didn't pay attention to the other parts.

  if the compilation is succesful, and no changes are needed
  to the Makefile if a new `cmd/other-nice-command/main.go`
  is created. "Write once, use forever".
- `mk/docker.mk`: Generic rules to build and push container images.
  It has an interesting set of variables that combined properly
  allow us a lot of customization and create new rules on them.
- `mk/db.mk`: Rules to start/stop a postgres container database
  and clean the data volume. It adds some other rules to quickly
  invoke the migration and seed the database with random values.
- `mk/plantuml.mk`: Invoke plantuml to generate diagrams in SVG
  format from the `docs/*.puml` files. It requires that plantuml
  is installed in your environment. Add a generic rule for it, so
  that no changes are needed when new `.puml` file is added to the
  `doc/` directory.

@Miscelanea

- `mk/printvars.mk`: It helps to fix wrong behaviors so it print out
  the variable name and the value not evaluated, so we can realize
  what final values could be applied and how should the values be
  expanded.
- `mk/alias.mk`: It provides compatibility with the initial set of
  rules in the Makefile, so that it would avoid impact during a

>   transition period to the new set of rules.

Reviewer 2: our project is so new, i think we should just drop the aliases alltogether, but i'll bring this up in our next standup :)

Author: thanks! it does not make many sense the `mk/alias.mk` then, but keeping it until the next standup.
